Added Auto discovery of volumes and create SC wizard

gnehapk opened the pull request:
Completed -
- Auto Detect Volume Flow for Local Storage Operator View
- Create Storage Cluster Wizard for ocs installation on attached devices
- Total Capacity available on the discovered nodes

![Screenshot from 2020-07-20 13-04-44](https://user-images.githubusercontent.com/5517376/87898846-4af3a300-ca8a-11ea-8a49-9e62f79dbb27.png)
![Screenshot from 2020-07-20 13-05-10](https://user-images.githubusercontent.com/5517376/87898852-4fb85700-ca8a-11ea-948c-65a8581ab09c.png)
![Screenshot from 2020-07-20 13-07-06](https://user-images.githubusercontent.com/5517376/87898854-521ab100-ca8a-11ea-8ac3-0e09a5bc2b6d.png)

Changes to be included in next PR -
- Donut chart using the total capacity calculated
- Update the selected capacity based on filter
- Do changes to incorporate CLI scenario in all UI views
- Hide non OCS specific options from LVS and ADV

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/wizard-pages/auto-detect-volume.tsx:
```diff
@@ -0,0 +1,24 @@
+import '../../attached-devices.scss';
+
+export const AutoDetectVolume: React.FC<AutoDetectVolumeProps> = ({ state, dispatch }) => {
+  return (
```
Remove return

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/install-lso-sc.tsx:
```diff
@@ -49,21 +49,36 @@ const makeOCSRequest = (
+  //   }
+
+  //   return () => {
+  //     stopWatchK8sResource(scResource);
```
Need to figure out a way to stop watching resource here as `stopWatchK8sResource` is used for class based component.

gnehapk commented:
/assign @cloudbehl @bipuladh

cloudbehl commented:
Please see the comments with each screenshot.


1. We need to `edit YAML` for the generic feature. 
2. Name for the CR is required & the user should be able to create multiple. How would I create new discoveries?
3. [Ask Backend]What if there are 500 nodes are we going to pass names for all nodes?
4. [UX]I think the location doesn't make sense on this page rather we should show status.
5. Why are we not showing the master nodes? Why can't user select master nodes
6. Too much space between selected nodes & node list.
<img width="1680" alt="Screenshot 2020-07-20 at 5 22 02 PM" src="https://user-images.githubusercontent.com/6695156/87934963-c87cda80-caad-11ea-94aa-75b7d08606e9.png">

1. The filter is not working properly. if you filter and select & deselect nodes the create button is getting disabled.

<img width="1680" alt="Screenshot 2020-07-20 at 5 36 36 PM" src="https://user-images.githubusercontent.com/6695156/87936087-de8b9a80-caaf-11ea-958e-c6e5bb085889.png">

gnehapk commented:
> Please see the comments with each screenshot.
> 
> 1. We need to `edit YAML` for the generic feature.

As discussed, will do once I am done with all the priority changes, as this low priority for now.
> 2. Name for the CR is required & the user should be able to create multiple. How would I create new discoveries?

   In 4.6, there will be just single instance of Auto Discovery CR. Hence, name not needed.
> 3. [Ask Backend]What if there are 500 nodes are we going to pass names for all nodes?

   For now, UI is passing all the selected nodes. But a good enhancement.
> 4. [UX]I think the location doesn't make sense on this page rather we should show status.

   Location might be helpful in case of aws with attached devices. Need to check with Yuval. 
> 5. Why are we not showing the master nodes? Why can't user select master nodes

The scope for this release is worker or schedule-able nodes.
> 6. Too much space between selected nodes & node list.
> 
> <img alt="Screenshot 2020-07-20 at 5 22 02 PM" width="1680" src="https://user-images.githubusercontent.com/6695156/87934963-c87cda80-caad-11ea-94aa-75b7d08606e9.png">
> 
Fixed
> 1. The filter is not working properly. if you filter and select & deselect nodes the create button is getting disabled.
> 
> <img alt="Screenshot 2020-07-20 at 5 36 36 PM" width="1680" src="https://user-images.githubusercontent.com/6695156/87936087-de8b9a80-caaf-11ea-958e-c6e5bb085889.png">

Fixed.

@cloudbehl Updated the PR. PTAL.

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/create-sc.tsx:
```diff
@@ -0,0 +1,263 @@
+import { CreateOCS } from '../install-lso-sc';
+import '../attached-devices.scss';
+
+enum CreateStepsSC {
```
Move this to some constants file.

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/create-sc.tsx:
```diff
@@ -0,0 +1,263 @@
+          {({ activeStep, onNext, onBack, onClose }) => {
+            if (activeStep.id !== CreateStepsSC.STORAGECLUSTER) {
+              return (
+                <>
```
Not needed. Whenever there is an error, the `Alert` element will get displayed and for that no action bar is needed, if I understand your comment here.

gnehapk replied:
![Screenshot from 2020-07-23 13-15-24](https://user-images.githubusercontent.com/5517376/88251731-b2a62a00-cce6-11ea-93fa-7fba8c52f883.png)

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/create-sc.tsx:
```diff
@@ -0,0 +1,263 @@
+    state.allNodeNamesOnADV,
+  ]);
+  // displaying node names for discovery step
+  React.useEffect(() => {
+    if ((nodeLoadError || nodeData.length === 0) && nodeLoaded) {
+      dispatch({ type: 'setAllNodeNamesOnADV', value: [] });
+    } else if (nodeLoaded) {
+      const names = nodeData.filter((node) => !hasTaints(node)).map((node) => getName(node));
+      dispatch({ type: 'setAllNodeNamesOnADV', value: names });
+    }
+  }, [nodeData, nodeLoaded, nodeLoadError]);
```
Yes, we need as its setting `allNodeNamesOnADV` state after filtering out the tainted nodes and this state is being used at various places. Since I am using the same logic in `create-sc.tsx` and `auto-detect-volume.tsx` both places, I will move this logic to just here `auto-detect-volume-inner.tsx`from where it can be used at these both places. Thanks for pointing it out.

gnehapk commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/create-sc/state.ts:
```diff
@@ -0,0 +1,166 @@
+export const diskModeDropdownItems = {
```
Already added types for all actions here - https://github.com/openshift/console/pull/6018/files#diff-6b6c11d422832c0498b4850e7ca3f832R8. Is this still needed?

bipuladh commented on frontend/packages/ceph-storage-plugin/src/components/ocs-install/attached-devices/install.tsx:
```diff
@@ -3,26 +3,28 @@ import { match as RouterMatch } from 'react-router';
-  const [scData, scLoaded, scLoadError] = useK8sWatchResource<StorageClassResourceKind[]>(
-    scResource,
-  );
+  const [isNewSCToBeCreated, setIsNewSCToBeCreated] = React.useState<boolean>(false);
```
```suggestion
  const [isNewSCToBeCreated, setIsNewSCToBeCreated] = React.useState(false);
```

bipuladh commented on frontend/packages/local-storage-operator-plugin/src/components/auto-detect-volume/auto-detect-volume.tsx:
```diff
@@ -0,0 +1,106 @@
+import { LocalVolumeDiscovery as AutoDetectVolumeModel } from '../../models';
+import { initialState, reducer } from './state';
+import { DISCOVERY_CR_NAME, LOCAL_STORAGE_NAMESPACE } from '../../constants';
+import { fetchK8s } from '@console/internal/graphql/client';
```
This goes before relative imports.
